query_server/query/src/extension/logical/optimizer_rule/implicit_type_conversion.rs

```diff
@@ -0,0 +1,746 @@
+                projection,
+                projected_schema,
+                filters,
+                fetch,
+            }) => {
+                let rewrite_filters = filters
+                    .clone()
+                    .into_iter()
+                    .map(|e| e.rewrite(&mut rewriter))
+                    .collect::<Result<Vec<_>>>()?;
+                Ok(LogicalPlan::TableScan(TableScan {
+                    table_name: table_name.clone(),
+                    source: source.clone(),
+                    projection: projection.clone(),
+                    projected_schema: projected_schema.clone(),
+                    filters: rewrite_filters,
+                    fetch: *fetch,
+                }))
+            }
+            LogicalPlan::Projection { .. }
+            | LogicalPlan::Window { .. }
+            | LogicalPlan::Aggregate { .. }
+            | LogicalPlan::Repartition { .. }
+            | LogicalPlan::CreateExternalTable { .. }
+            | LogicalPlan::Extension { .. }
+            | LogicalPlan::Sort { .. }
+            | LogicalPlan::Explain { .. }
+            | LogicalPlan::Limit { .. }
+            | LogicalPlan::Union { .. }
+            | LogicalPlan::Join { .. }
+            | LogicalPlan::CrossJoin { .. }
+            | LogicalPlan::CreateMemoryTable { .. }
+            | LogicalPlan::DropTable { .. }
+            | LogicalPlan::Values { .. }
+            | LogicalPlan::Distinct { .. }
+            | LogicalPlan::Analyze { .. } => {
+                let inputs = plan.inputs();
+                let new_inputs = inputs
+                    .iter()
+                    .map(|plan| self.optimize(plan, optimizer_config))
+                    .collect::<Result<Vec<_>>>()?;
+
+                let expr = plan
+                    .expressions()
+                    .into_iter()
+                    .map(|e| e.rewrite(&mut rewriter))
+                    .collect::<Result<Vec<_>>>()?;
+
+                utils::from_plan(plan, &expr, &new_inputs)
+            }
+
+            LogicalPlan::Subquery(_)
+            | LogicalPlan::SubqueryAlias(_)
+            | LogicalPlan::CreateView(_)
+            | LogicalPlan::CreateCatalogSchema(_)
+            | LogicalPlan::CreateCatalog(_)
+            | LogicalPlan::EmptyRelation { .. } => Ok(plan.clone()),
+        }
+    }
+
+    fn name(&self) -> &str {
+        "implicit_type_conversion"
+    }
+}
+
+struct DataTypeRewriter<'a> {
+    /// input schemas
+    schemas: Vec<&'a DFSchemaRef>,
+}
+
+impl<'a> DataTypeRewriter<'a> {
+    /// If expr is Column, get its data type; otherwise, return None
+    fn extract_column_data_type(&self, expr: &Expr) -> Option<DataType> {
+        if let Expr::Column(_) = expr {
+            for schema in &self.schemas {
+                if let Ok(v) = expr.get_type(schema) {
+                    return Some(v);
+                }
+            }
+        }
+
+        None
+    }
+
+    fn convert_data_type_if_necessary<'b>(
```

Comment:
I have an idea to avoid these clones.

```rust
fn convert_data_type_if_necessary(
      &self,
      mut left: Box<Expr>,
      mut right: Box<Expr>,
  ) -> Result<(Box<Expr>, Box<Expr>)> {
      let left_type = self.extract_column_data_type(&left);
      let right_type = self.extract_column_data_type(&right);

      // Ensure that the left side of the op is column
      let mut reverse = false;
      let left_type = match (&left_type, &right_type) {
          (Some(v), None) => v,
          (None, Some(v)) => {
              reverse = true;
              mem::swap(&mut left, &mut right);
              v
          }
          _ => return Ok((left, right)),
      };

      // Only processing of column op literal
      match (left.as_ref(), right.as_ref()) {
          // Convert the data on the right of op to the data type corresponding to the left column
          (Expr::Column(col), Expr::Literal(value)) => {
              let casted_right = Self::cast_scalar_value(&value, left_type)?;
              debug!(
                  "DataTypeRewriter convert type, origin_left:{:?}, type:{}, right:{:?}, casted_right:{:?}",
                  col, left_type, value, casted_right
              );
              if casted_right.is_null() {
                  return Err(DataFusionError::Plan(format!(
                      "column:{:?} value:{:?} is invalid",
                      col, value
                  )));
              }
              if reverse {
                  Ok((Box::new(Expr::Literal(casted_right)), left))
              } else {
                  Ok((left, Box::new(Expr::Literal(casted_right))))
              }
          }
          _ => Ok((left, right)),
      }
  }
```

Reply:
But `ExprRewriter::mutate()` must be mutated using that code.

```rust
...
fn mutate(&mut self, expr: Expr) -> Result<Expr> {
...
    Expr::InList {
        expr,
        list,
        negated,
    } => {
        let mut list_expr = Vec::with_capacity(list.len());
        for e in list {
            let (_, expr_conversion) = self.convert_data_type_if_necessary(expr.clone(), Box::new(e))?;
            list_expr.push(*expr_conversion);
        }
        Expr::InList {
            expr,
            list: list_expr,
            negated,
        }
    }
    ...
}
```